DO NOT MERGE: 28.0.0 release notes (#15173)
apache/druid: docs/do-not-merge.md

Support for Hadoop 2 has been removed.
Migrate to SQL-based ingestion or native ingestion if you are using Hadoop 2.x for ingestion today. If migrating to Druid's built-in ingestion is not possible, you must upgrade your Hadoop infrastructure to 3.x+ before upgrading to Druid 28.0.0.

[#14763](https://github.com/apache/druid/pull/14763)

# Additional features and improvements

## SQL-based ingestion

### New ingestion mode for arrays

The MSQ task engine now includes an `arrayIngestMode` that determines how arrays are treated for ingestion. Previously, arrays were ingested as multi-value dimensions (MVDs). They can now be ingested as `ARRAY<STRING>` or numeric arrays instead if the ingest mode is set to `array`. 

The default mode is for MVDs, which is unchanged from previous releases.

[#15093](https://github.com/apache/druid/pull/15093)

### UNNEST functionality

You can now use UNNEST with MSQ queries. This lets you flatten and explode data during batch ingestion. For more information, see the [UNNEST tutorial](https://druid.apache.org/docs/latest/tutorials/tutorial-unnest-arrays/) and [UNNEST documentation](https://druid.apache.org/docs/latest/querying/sql/#unnest).

> 

Reviewer: Should also put a note that the context flag "enableUnnest" is not needed anymore to run queries involving unnest

Author: Updated

[#15058](https://github.com/apache/druid/pull/15058)

### Strict booleans

`druid.expressions.useStrictBooleans` is now enabled by default.
Druid now handles booleans strictly using `1` (true) or `0` (false).
Previously, true and false could be represented either as `true` and `false`, respectively, as well as `1` and `0`.

If you don't explicitly configure this property in `runtime.properties`, clusters now use LONG types for any ingested boolean values, and in the output of boolean functions for transformations and query time operations.

This change may impact your query results. For more information, see [SQL compatibility in the upgrade notes](#sql-compatibility-1).

[#14734](https://github.com/apache/druid/pull/14734)

### NULL handling

`druid.generic.useDefaultValueForNull` is now disabled by default.
Druid now differentiates between empty records, such as `' '`, and null records.
Previously, Druid might treat empty records as empty or null.

This change may impact your query results. For more information, see [SQL compatibility in the upgrade notes](#sql-compatibility-1).

[#14792](https://github.com/apache/druid/pull/14792)

## SQL-based ingestion

### Azure connector

> Added support for Microsoft's Azure storage type.

Reviewer 2: ```suggestion
Added support for Microsoft's Azure Blob Storage.
```

Author: Updated

> You can now use fault tolerance and durable storage with Microsoft Azure's blob storage.

Reviewer 2: ```suggestion
You can now use fault tolerance and durable storage with Microsoft Azure's Blob Storage.
```

For more information, see [Durable storage](https://druid.apache.org/docs/latest/multi-stage-query/reference#durable-storage).

[#14660](https://github.com/apache/druid/pull/14660)

### New ingestion mode for arrays

> The MSQ task engine now includes an `arrayIngestMode` that determines how arrays are treated for ingestion. Previously, arrays were ingested as multi-value dimensions (MVDs). They can now be ingested as `ARRAY<STRING>` or numeric arrays instead if the ingest mode is set to `array`.

Reviewer 2: ```suggestion
The MSQ task engine now includes an `arrayIngestMode` that determines how string arrays are stored. MSQ till Druid 27 supported string arrays, and they were ingested as multi-value dimensions (MVDs). This is incorrect behavior,  and now they can now be ingested as string arrays as intended. This is incompatible with the existing behavior so please refer to the migration docs to learn more on how to migrate to the new behaviour.
```

Reviewer 2: Ingestion mode primarily affects string arrays, and the release note should reflect that. Also, we should probably note that the default behavior is `mvd` instead of `array` (new behavior) and numeric arrays can only be ingested if `array` is set in the query context.

Author: @LakshSingla If MSQ supported string arrays until Druid 27, what happened in Druid 27? Did we stop supporting string arrays in Druid 27?

Reviewer 3: Druid 27 and older incorrectly stored them as multi-value STRING columns instead of ARRAY<STRING>, without requiring wrapping the string array in `ARRAY_TO_MV` which is what should have been done. The docs I am working on should help clear this up

Author: Updated

* `druid.generic.useThreeValueLogicForNativeFilters = true`
* `druid.expressions.useStrictBooleans = true`
* `druid.generic.useDefaultValueForNull = false`

[#15058](https://github.com/apache/druid/pull/15058)

### Strict booleans

`druid.expressions.useStrictBooleans` is now enabled by default.
Druid now handles booleans strictly using `1` (true) or `0` (false).
Previously, true and false could be represented either as `true` and `false` as well as `1` and `0`, respectively.
In addition, Druid now returns a null value for Boolean comparisons like `True && NULL`.

If you don't explicitly configure this property in `runtime.properties`, clusters now use LONG types for any ingested boolean values and in the output of boolean functions for transformations and query time operations.

This change may impact your query results. For more information, see [SQL compatibility in the upgrade notes](#sql-compatibility-1).

[#14734](https://github.com/apache/druid/pull/14734)

### NULL handling

`druid.generic.useDefaultValueForNull` is now disabled by default.
Druid now differentiates between empty records and null records.
Previously, Druid might treat empty records as empty or null.

> This change may impact your query results. For more information, see [SQL compatibility in the upgrade notes](#sql-compatibility-1).

Author: I am going to remove this sentence since we provide additional details in the upgrade notes.

[#14792](https://github.com/apache/druid/pull/14792)

## SQL planner improvements

> Druid uses Apache Calcite for SQL planning and optimization. Starting in Druid 28.0.0, the Calcite version has been upgraded from 1.21 to 1.35. This upgrade brings in many bug fixes in SQL planning from Calcite. As part of the upgrade, the behavior of type inference for [dynamic parameters](#dynamic-parameters) and the [recommended syntax for UNNEST](#new-syntax-for-sql-unnest) have changed.

Reviewer 3: this reads kind of funny, the updated recommendation for unnest isn't directly related to the calcite upgrade afaik, rather it is just less likely to run into problems with complicated queries if you use the more verbose cross join syntax

Author: @clintropolis Since the updated recommendation for unnest isn't directly related to the calcite upgrade, I think we should move to "SQL UNNEST and ingestion flattening".

### Dynamic parameters

The behavior of type inference for dynamic parameters has changed. To avoid any type interference issues, explicitly `CAST` all dynamic parameters as a specific data type in SQL queries. For example, use:

```sql
SELECT (1 * CAST (? as DOUBLE))/2 as tmp
```

Do not use:

```sql
SELECT (1 * ?)/2 as tmp
```

> ### New syntax for SQL UNNEST

Author: Changing the heading to "Recommended syntax for SQL UNNEST" since the syntax is not actually new.

* You can now select a format when downloading data [#14794](https://github.com/apache/druid/pull/14794)
* Improved the clarity of cluster default rules in the retention dialog [#14793](https://github.com/apache/druid/pull/14793)
* The web console now detects inline queries in the query text and lets you run them individually [#14810](https://github.com/apache/druid/pull/14801)
* You can now reset specific partition offsets for a supervisor [#14863](https://github.com/apache/druid/pull/14863)

## Multi-stage query

### Support for multiple result formats

Query from deep storage now supports multiple result formats.
Previously, the `/druid/v2/sql/statements/` endpoint only supported results in the `object` format. Now, results can be written in any format specified in the `resultFormat` parameter.
For more information on result parameters supported by the Druid SQL API, see [Responses](https://druid.apache.org/docs/latest/api-reference/sql-api#responses).

[#14571](https://github.com/apache/druid/pull/14571)

### Broadened access for queries from deep storage

> Users with the `STATE` permission can interact with status APIs for queries from deep storage. Previously, only the user who submitted the query could use those APIs. This enables the web console to monitor the running status of the queries. Users with the `STATE` permission can access the query results.

Author: @clintropolis Originally, I had PR 14571 and PR 14944 listed under the deep storage section. @2bethere proposed that we change the outline. @abhagraw @LakshSingla any thoughts?

> Users with the `STATE` permission can interact with status APIs for queries from deep storage. Previously, only the user who submitted the query could use those APIs. This enables the web console to monitor the running status of the queries. Users with the `STATE` permission can access the query results.

Reviewer 3: i personally think we should combine stuff as much as possible - i kind of felt like i was re-reading stuff i had already read in a few places, which is why i brought this up

> Users with the `STATE` permission can interact with status APIs for queries from deep storage. Previously, only the user who submitted the query could use those APIs. This enables the web console to monitor the running status of the queries. Users with the `STATE` permission can access the query results.

Author: Moved PR 14571 and PR 14944 to the Query from deep storage section.

Druid now stops loading and moving segments as soon as they are marked as unused. This prevents Historical processes from spending time on superfluous loads of segments that will be unloaded later. You can mark segments as unused by a drop rule, overshadowing, or by calling [the Data management API](https://druid.apache.org/docs/latest/api-reference/data-management-api).

[#14644](https://github.com/apache/druid/pull/14644)

## Compaction

### Task locks for append and replace batch ingestion jobs

Append batch ingestion jobs can now share locks. This allows you to run multiple append batch ingestion jobs against the same time internal. Replace batch ingestion jobs still require an exclusive lock. This means you can run multiple append batch ingestion jobs and one replace batch ingestion job for a given interval.

[#14407](https://github.com/apache/druid/pull/14407)

### Streaming ingestion with concurrent replace

Streaming jobs reading from Kafka and Kinesis with `APPEND` locks can now ingest concurrently with compaction running with `REPLACE` locks. The segment granularity of the streaming job must be equal to or finer than that of the concurrent replace job.

> [#15039](https://github.com/apache/druid/pull/15039)

Reviewer 3: should this (and web-console section) all be combined into single section about concurrent append and replace?

Author: I think it makes sense to add PR 14407 and PR 15039 as subsections to the main Concurrent append and replace section. I'd like to keep concurrent append and replace UI improvements with the rest of the web console improvements.

[#14456](https://github.com/apache/druid/pull/14456)

### Ingestion status

Ingestion reports now include a `segmentLoadStatus` object that provides information related to the ingestion, such as duration and total segments.

[#14322](https://github.com/apache/druid/pull/14322)

### SQL-based ingestion

#### Ability to ingest ARRAY types

SQL-based ingestion now supports storing ARRAY typed values in [ARRAY typed columns](https://druid.apache.org/docs/latest/querying/arrays) as well as storing both VARCHAR and numeric typed arrays.
Previously, the MSQ task engine stored ARRAY typed values as [multi-value dimensions](https://druid.apache.org/docs/latest/querying/multi-value-dimensions) instead of ARRAY typed columns.

The MSQ task engine now includes the `arrayIngestMode` query context parameter, which controls how
`ARRAY` types are stored in Druid segments.
Set the `arrayIngestMode` query context parameter to `array` to ingest ARRAY types.

In Druid 28.0.0, the default mode for `arrayIngestMode` is `mvd` for backwards compatibility, which only supports VARCHAR typed arrays and stores them as multi-value dimensions. This default is subject to change in future releases.

> Note that this improvement is incompatible with previous Druid versions. For information on how to migrate to the new behavior, see the [Ingestion options for ARRAY typed columns in the upgrade notes](#ingestion-options-for-array-typed-columns).

Reviewer 3: this isn't strictly true, the ARRAY columns were introduced in Druid 27, the only problem would be if someone was trying to use the new `"arrayIngestMode":"array"` and had not upgraded all of the nodes the task run on resulting in a mix of behavior

Author: Removed the sentence about incompatibility with previous Druid versions.

## Extensions

### Ingestion task payloads for Kubernetes

You can now provide compressed task payloads larger than 128 KB when you run MiddleManager-less ingestion jobs.

[#14887](https://github.com/apache/druid/pull/14887)

### Prometheus emitter

The Prometheus emitter now supports a new optional configuration parameter, `druid.emitter.prometheus.extraLabels`.
This addition offers the flexibility to add arbitrary extra labels to Prometheus metrics, providing more granular control in managing and identifying data across multiple Druid clusters or other dimensions.
For more information, see [Prometheus emitter extension](https://druid.apache.org/docs/latest/development/extensions-contrib/prometheus).

[#14728](https://github.com/apache/druid/pull/14728)

## Documentation improvements

We've moved Jupyter notebooks that guide you through query, ingestion, and data management with Apache Druid to the new [Learn Druid](https://github.com/implydata/learn-druid) repository.
The repository also contains a Docker Compose file to get you up and running with a learning lab.

[#15136](https://github.com/apache/druid/pull/15136)

# Upgrade notes and incompatible changes

> ## Upgrade notes

Reviewer 3: I think we should do something a bit different than previous releases before we finally publish these release notes, where this section is about upgrading from Druid 27, and then we also should include the upgrade notices from the previous few releases as well, just copied in from earlier release notes. So that way we have all the information in one place to cover like a year worth of upgrades. I don't think this needs to be added yet, we can wait until the last minute before publishing, but I think it would be helpful since operators don't always upgrade every release

When you set `arrayIngestMode` to `none`, Druid throws an exception when trying to store any type of arrays.

For more information on how to ingest `ARRAY` typed columns with SQL-based ingestion, see [SQL data types](https://druid.apache.org/docs/latest/querying/sql-data-types#arrays) and [Array columns](https://druid.apache.org/docs/latest/querying/arrays).

## Incompatible changes

### Removed Hadoop 2

Support for Hadoop 2 has been removed.
Migrate to SQL-based ingestion or JSON-based batch ingestion if you are using Hadoop 2.x for ingestion today.
If migrating to Druid's built-in ingestion is not possible, you must upgrade your Hadoop infrastructure to 3.x+ before upgrading to Druid 28.0.0.

[#14763](https://github.com/apache/druid/pull/14763)

### Removed GroupBy v1 

The GroupBy v1 engine has been removed. Use the GroupBy v2 engine instead.  
If you are using GroupBy v1 in JSON-based batch queries, you must change your queries before upgrading. Otherwise, your queries will fail.

> For more information, see [GroupBy queries](https://druid.apache.org/docs/latest/querying/groupbyquery.html).

Reviewer 3: hm, i don't think the queries fail, just the option is ignored completely and grouping always uses v2, but i can double check...

Author: @clintropolis Can you please double check? I can update the message based on your findings.

Reviewer 3: yeah, it is just ignored

Author: Updated

[#14866](https://github.com/apache/druid/pull/14866)

### Removed Coordinator dynamic configs

The `decommissioningMaxPercentOfMaxSegmentsToMove` config has been removed.
The use case for this config is handled by smart segment loading now, which is enabled by default.

[#14923](https://github.com/apache/druid/pull/14923)

### Removed `cachingCost` strategy

The `cachingCost` strategy for segment loading has been removed.
Use `cost` instead, which has the same benefits as `cachingCost`.

If you have `cachingCost` set, the system ignores this setting and automatically uses `cost`.

[#14798](https://github.com/apache/druid/pull/14798)

### Removed `InsertCannotOrderByDescending`

The deprecated MSQ fault `InsertCannotOrderByDescending` has been removed.

[#14588](https://github.com/apache/druid/pull/14588)

### Removed the backward compatibility code for the Handoff API

The backward compatibility code for the Handoff API in `CoordinatorBasedSegmentHandoffNotifier` has been removed.

[#14652](https://github.com/apache/druid/pull/14652)

> 

Reviewer 3: heh, should we call out if you are upgrading from a Druid version older than 0.14 you're going to have a bad time? (probably not necessary, but it might be nicer to make it super clear that this is a very old backwards compat thing so people don't worry)

Author: @clintropolis If we do add a note about upgrading from a Druid version older than 0.14, what should the message say? What are we warning the users about?

Reviewer 3: I'm not sure we even need to mention it? Otherwise, i guess we could mention something about upgrading to a newer version of Druid before upgrading to 28

Author: Added a note about upgrading from Druid 0.14.0

[#14763](https://github.com/apache/druid/pull/14763)

## JSON and auto column indexer

The `json` type is now equivalent to using `auto` in native JSON-based ingestion dimension specs. Upgrade your ingestion specs to `json` to take advantage of the features and functionality of `auto`, including the following:

- Type specializations including ARRAY typed columns
- Better support for nested arrays of strings, longs, and doubles
- Smarter index utilization

`json` type columns created with Druid 28.0.0 are not backwards compatible with Druid versions older than 26.0.0.
If you upgrade from one of these versions, you can continue to write nested columns in a backwards compatible format (version 4).

For more information, see [Nested column format in the upgrade notes](#nested-column-format).

[#14955](https://github.com/apache/druid/pull/14955)
[#14456](https://github.com/apache/druid/pull/14456)

# Additional features and improvements

## SQL compatibility

Druid continues to make SQL query execution more consistent with how standard SQL behaves. However, there are feature flags available to restore the old behaviour if needed.

### Three-valued logic

> Druid native filters now correctly observe SQL [three-valued logic](https://en.wikipedia.org/wiki/Three-valued_logic#SQL), `true`, `false`, `unknown`, instead of Druid's classic two-state logic when you set the following configuration values:

Author: Updated

* `druid.generic.useThreeValueLogicForNativeFilters = true`
* `druid.expressions.useStrictBooleans = true`
* `druid.generic.useDefaultValueForNull = false`

[#15058](https://github.com/apache/druid/pull/15058)

### Strict booleans

`druid.expressions.useStrictBooleans` is now enabled by default.
Druid now handles booleans strictly using `1` (true) or `0` (false).
Previously, true and false could be represented either as `true` and `false`, respectively, as well as `1` and `0`.

If you don't explicitly configure this property in `runtime.properties`, clusters now use LONG types for any ingested boolean values, and in the output of boolean functions for transformations and query time operations.

This change may impact your query results. For more information, see [SQL compatibility in the upgrade notes](#sql-compatibility-1).

[#14734](https://github.com/apache/druid/pull/14734)

### NULL handling

`druid.generic.useDefaultValueForNull` is now disabled by default.

> Druid now differentiates between empty records, such as `' '`, and null records.

Author: Fixed

Previously, Druid might treat empty records as empty or null.

This change may impact your query results. For more information, see [SQL compatibility in the upgrade notes](#sql-compatibility-1).

[#14792](https://github.com/apache/druid/pull/14792)

## SQL-based ingestion

### Ability to ingest ARRAY types

SQL-based ingestion now supports storing ARRAY typed values in [ARRAY typed columns](https://druid.apache.org/docs/latest/querying/arrays) as well as storing both VARCHAR and numeric typed arrays.

> Previously, the MSQ task engine incorrectly stored ARRAY typed values as [multi-value dimensions](https://druid.apache.org/docs/latest/querying/multi-value-dimensions) instead of ARRAY typed columns.

Author: Removed "incorrectly"

The MSQ task engine now includes the `arrayIngestMode` query context parameter, which controls how
`ARRAY` types are stored in Druid segments.
Set the `arrayIngestMode` query context parameter to `array` to ingest ARRAY types.

> In Druid 28.0.0, the default mode for `arrayIngestMode` is `mvd` for backwards compatibility, which only supports VARCHAR typed arrays and stores them as multi-value dimensions. This default is subject to removal in future releases.

Author: Good point. Updated the sentence to read "This default is subject to change in future releases."

* You can now specify the following parameters for a `kill` task:
  * `batchSize`: The maximum number of segments to delete in one `kill` batch [#14642](https://github.com/apache/druid/pull/14642)
  * `limit`: The maximum number of segments for a `kill` task to delete [#14662](https://github.com/apache/druid/pull/14662)
* You can now speed up `kill` tasks by batch deleting multiple segments stored in S3 [#14131](https://github.com/apache/druid/pull/14131)
* Kill tasks that delete unused segments now publish a task report containing kill stats such as `numSegmentsKilled`, `numBatchesProcessed`, and `numSegmentsMarkedAsUnused` [#15023](https://github.com/apache/druid/pull/15023)
* `IndexerSQLMetadataStorageCoordinator` now uses the JDBI `PreparedBatch` instead of issuing single update statements inside a transaction to mitigate scaling challenges [#14639](https://github.com/apache/druid/pull/14639)

## Web console

### Added UI support for segment loading query context parameter

> The web console now supports the `waitUntilSegmentsLoad` query context parameter.

Reviewer 4: It's odd to have the "now" in this one and the concurrent compaction one since these UI changes coincide with the introduction of those features

Author: Removed "now" from both

![UI for waitUntilSegmentsLoad context parameter](image.png)

[#15110](https://github.com/apache/druid/pull/15110)

> ### Added concurrent compaction switches

Reviewer 4: ```suggestion
### Added concurrent append and replace switches
```

Concurrent compaction -> concurrent append and replace in the body too

Author: Updated

The web console now includes concurrent compaction switches.

The following screenshot shows the concurrent compaction switch in the classic batch ingestion wizard:
![Classic batch ingestion wizard](image-1.png)

The following screenshot shows the concurrent compaction switch in the compaction configuration UI:
![Compaction configuration UI](image-2.png)

[#15114](https://github.com/apache/druid/pull/15114)

### Other web console improvements

> * You can select to allow concurrent APPEND tasks in the web console [#15114](https://github.com/apache/druid/pull/15114)

Reviewer 4: duplicate of the above section

Author: Removed

* The web console now shows the execution dialog for `query_controller` tasks in the task view instead of the generic raw task details dialog. You can still access the raw task details from the ellipsis (...) menu [#14930)](https://github.com/apache/druid/pull/14930)
* You can now select a horizontal range in the web console time chart to modify the current WHERE clause [#14929](https://github.com/apache/druid/pull/14929)
* You can now set dynamic query parameters in the web console [#14921](https://github.com/apache/druid/pull/14921)
* You can now edit the Coordinator dynamic coonfiguration in the web console [#14791](https://github.com/apache/druid/pull/14791)
* You can now prettify SQL queries and use flatten with a Kafka input format [#14906](https://github.com/apache/druid/pull/14906)
* A warning now appears when a CSV or TSV sample contains newlines that Druid does not accept [#14783](https://github.com/apache/druid/pull/14783)
* You can now select a format when downloading data [#14794](https://github.com/apache/druid/pull/14794)
* Improved the clarity of cluster default rules in the retention dialog [#14793](https://github.com/apache/druid/pull/14793)

> * Enabled Kafka multi-topic ingestion from the data loader [#14833](https://github.com/apache/druid/pull/14833)

Reviewer 4: This maybe deserves its own H3 since the feature it's supporting has its H3 like concurrent append and replace does.

Author: Moved id 14833 to H3

The MSQ task engine now includes the `arrayIngestMode` query context parameter, which controls how
`ARRAY` types are stored in Druid segments.
Set the `arrayIngestMode` query context parameter to `array` to ingest ARRAY types.

In Druid 28.0.0, the default mode for `arrayIngestMode` is `mvd` for backwards compatibility, which only supports VARCHAR typed arrays and stores them as multi-value dimensions. This default is subject to change in future releases.

Note that this improvement is incompatible with previous Druid versions. For information on how to migrate to the new behavior, see the [Ingestion options for ARRAY typed columns in the upgrade notes](#ingestion-options-for-array-typed-columns).

For information on inserting, filtering, and grouping behavior for `ARRAY` typed columns, see [Array columns](https://druid.apache.org/docs/latest/querying/arrays).

[#15093](https://github.com/apache/druid/pull/15093)

### Ingestion status

Ingestion reports now include a `segmentLoadStatus` object that provides information related to the ingestion, such as duration and total segments.

[#14322](https://github.com/apache/druid/pull/14322)

> ### Azure connector

Reviewer 5: ```suggestion
### Azure Blob Storage support
```

Author: Updated
